### ISSUE TITLE
Inconsistent selection resulting from using `|`

### ISSUE THREAD

**Issue by lenormf** [mawww/kakoune#2394]
### Steps 

* create a multiple selection with each holding a mathematical operation (e.g. `10o <c-r># +1 <esc><a-x>`)
* use `|bc`

### Outcome 

The anchor/cursor of the selection are sometimes misplaced.

### Expected 

The output of the `bc` command correctly replaces the original selection.

**Comment by occivink** [comment]
This is related to the diff algorithm mentioned in #2037

**Comment by mawww** [comment]
Yep, this is the result of Kakoune diffing pipe input and pipe output in order to detect what  changes actually happened. I think we could tweak that behaviour by diffing on lines instead of character (which would match what we do on reload). This would fix the `bc` case (as we are replacing a line with another line), and would still be usefull in the motivating case for diff (piping to sort/indent/fmt...). Additionally it would likely solve performance problems (as diffing is O(N*D), with N the lenght of strings and D the amount of differences).

**Comment by Screwtapello** [comment]
> Yep, this is the result of Kakoune diffing pipe input and pipe output in order to detect what changes actually happened.

But why does Kakoune need to know what changes actually happened? If bytes 16 to 123 of the document are selected and piped to an external command that returns 27 bytes, then I would expect Kakoune to have bytes 16 to 43 (16 + 27) selected, regardless of how similar or different the new buffer is to the old, or how many lines are added or removed.

**Comment by daboross** [comment]
> But why does Kakoune need to know what changes actually happened?

I'd assume it's for the possibility of other clients with cursors in the middle of the expression.

**Comment by mawww** [comment]
@Screwtapello AFAIR its mostly for undo/redo, say if you sort a set of lines by piping to `sort`, having Kakoune detect what actually changed makes it much nicer to work with, especially on changes where you pipe the whole buffer (through a code formatter for example). Its also necessary so that other windows on the same text do not lose their selections.
